File: rileylink/src/main/java/info/nightscout/androidaps/plugins/pump/common/hw/medlink/defs/MedLinkCommandType.java
package info.nightscout.androidaps.plugins.pump.common.hw.medlink.defs;

import static java.nio.charset.StandardCharsets.UTF_8;

import java.util.Arrays;
import java.util.Iterator;

/**
 * Created by dirceu on 17/09/2020.
 */

public enum MedLinkCommandType {

    NoCommand(""),
    ReadCharacteristic("\"ReadCharacteristic\""),
    Notification("SetNotificationBlocking"),
    PumpModel("OK+CONN"),
    Connect("OK+CONN"),
    GetState("S"), //
    StopStartPump("A"),
    Bolus("X", true),
    BolusAmount("BOLUS", true),
    StartPump("START"),
    StopPump("STOP"),
    IsigHistory("I"),
    PreviousIsigHistory("J"),
    BGHistory("C"),
    PreviousBGHistory("T"),
    BolusHistory("H"),
    ActiveBasalProfile("E"),
    BaseProfile("F"),
    Calibrate("K", true),
    CalibrateFrequency("U"),
    CalibrateFrequencyArgument(""),
    CalibrateValue("CAL", true),
    BolusStatus("M"),
    SMBBolus("X", true),
    TBRBolus("X", true),
    PreviousBolusHistory("G",true)
    ;

    public final String code;
    public final boolean needActivePump;
    public Iterator<String> config;
    public Double insulinAmount = 0d;
    public Double bgValue = 0d;
    public Integer resourceId = null;

    public String getCommandDescription() {
        return this.name();
    }
//    MedLinkCommandType(MedLinkCommandType command, Double insulinAmount) {
//        this.code = command.code;
//        this.insulinAmount = insulinAmount;
//    }

    MedLinkCommandType(String command){ this(command,false);
    }

    MedLinkCommandType(String command, boolean needActivePump) {
        this.code = command;
        this.needActivePump = needActivePump;
    }

    private byte[] appendEndLine(StringBuilder buff){
        buff.append("\r").append("\n");
        return buff.toString().getBytes(UTF_8);
    }

    public byte[] getRaw() {
        if(this.config != null && this.config.hasNext()){
            StringBuilder buffer = new StringBuilder(config.next());
            return appendEndLine(buffer);
        }else
        if (this.insulinAmount > 0) {
            StringBuilder buff = new StringBuilder(this.code);
            buff.append(" ");
            if (this.insulinAmount < 10d) {
                buff.append(" ");
            }
            buff.append(this.insulinAmount.toString());

            return appendEndLine(buff);
        } else if(bgValue!=0d){
            StringBuilder buff = new StringBuilder(this.code);
            buff.append(" ");
            if (this.bgValue < 100d) {
                buff.append("0");
            }
            buff.append(this.bgValue.intValue());
            return appendEndLine(buff);
        } else if (this.code != null && !this.code.isEmpty()) {
            return appendEndLine(new StringBuilder(this.code));
        } else {
            return new byte[0];
        }
    }

    public boolean isSameCommand(byte[] command) {
        return command != null && Arrays.equals(this.getRaw(), command);
    }

    public boolean isSameCommand(String command) {
        return command != null && command.equals(this.code);
    }

    public boolean isSameCommand(MedLinkCommandType command) {
        return command != null && (isSameCommand(command.getRaw()) || isSameCommand(command.code));
    }
}

File: rileylink/src/main/java/info/nightscout/androidaps/plugins/pump/common/hw/medlink/defs/MedLinkCommandType.kt
package info.nightscout.androidaps.plugins.pump.common.hw.medlink.defs

import kotlin.jvm.JvmOverloads
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.defs.MedLinkCommandType
import java.lang.StringBuilder
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * Created by dirceu on 17/09/2020.
 */
enum class MedLinkCommandType @JvmOverloads constructor(command: String, needActivePump: Boolean = false) {

    NoCommand(""),
    ReadCharacteristic("\"ReadCharacteristic\""),
    Notification("SetNotificationBlocking"),
    PumpModel("OK+CONN"),
    Connect("OK+CONN"),
    GetState("S"),  //
    StopStartPump("A"),
    Bolus("X", true),
    BolusAmount("BOLUS", true),
    StartPump("START",true),
    StopPump("STOP",true),
    IsigHistory("I"),
    PreviousIsigHistory("J"),
    BGHistory("C"),
    PreviousBGHistory("T"),
    BolusHistory("H"),
    ActiveBasalProfile("E"),
    BaseProfile("F"),
    Calibrate("K", true),
    CalibrateFrequency("U"),
    CalibrateFrequencyArgument(""),
    CalibrateValue("CAL", true),
    BolusStatus("M"),
    SMBBolus("X", true),
    TBRBolus("X", true),
    PreviousBolusHistory("G", false);

    @JvmField val code: String?
    val needActivePump: Boolean
    var config: Iterator<String>? = null
    var insulinAmount = 0.0
    var bgValue = 0.0
    var resourceId: Int = 0
    fun getCommandDescription(): String {
        return name
    }

    private fun appendEndLine(buff: StringBuilder): ByteArray {
        buff.append("\r").append("\n")
        return buff.toString().toByteArray(StandardCharsets.UTF_8)
    }

    fun getRaw(argument: String = ""): ByteArray {
        // if (config != null && config!!.hasNext()) {
        //     val buffer = StringBuilder(config!!.next())
        //     appendEndLine(buffer)
        // } else
        return  if (this.code != null && this.code.isNotEmpty()) {
            appendEndLine(StringBuilder(this.code).append(argument))
        } else {
            ByteArray(0)
        }
    }

    fun isSameCommand(command: ByteArray?): Boolean {
        return command != null && Arrays.equals(getRaw(), command)
    }

    fun isSameCommand(command: String?): Boolean {
        return command != null && command == this.code
    }

    fun isSameCommand(command: MedLinkCommandType?): Boolean {
        return command != null && (isSameCommand(command.getRaw()) || isSameCommand(command.code))
    }

    //    MedLinkCommandType(MedLinkCommandType command, Double insulinAmount) {
    //        this.code = command.code;
    //        this.insulinAmount = insulinAmount;
    //    }
    init {
        this.code = command
        this.needActivePump = needActivePump
    }

    override fun toString(): String {
        return when {
            (this == BolusAmount) -> {
                super.toString() +
                    "bolusAmount: $insulinAmount"
            }
            (this == CalibrateFrequencyArgument) -> {
                super.toString() +
                    "calibrateArgument: $"

            }

            (this == CalibrateValue) -> {
                super.toString() +
                    "bgValue: $bgValue"

            }

            else                                 -> {
                super.toString()
            }
        }
    }
}